packages/SwingSet/docs/virtual-objects.md

```diff
@@ -0,0 +1,133 @@
+# Vat Secondary Storage
+
+The kernel currently provides two secondary storage mechanisms for the use of (application) code running in a vat:
+
+- Virtual objects
+- Persistent weak stores
+
+Each of these is accessed via a global made available to vat code.
+
+## Virtual Objects
+
+A virtual object is an object with a durable identity whose state is automatically and transparently backed up in secondary storage.  This means that when a given virtual object is not in active use, it need not occupy any memory within the executing vat.  Thus a vat can manage an arbitrarily large number of such objects without concern about running out of memory.
+
+A virtual object has a "kind", which defines what sort of behavior and state it will possess.  A kind is not exactly a data type, since it comes with a concrete implementation, but it indicates a family of objects that share a set of common behaviors and a common state template.
+
+A vat can define new kinds of virtual object by calling the `makeKind` function provided as a vat global:
+
+`makeKind(bodyMaker)`
+
+The return value from `makeKind` is a maker function which the vat can use to create instances of the newly defined object kind.
+
+The `bodyMaker` parameter is a user provided function that binds an in-memory instance of the virtual object with the virtual object's state, associating such instances with the virtual object's behavior.  It is called whenever a new virtual object instance is created, whenever such an instance is swapped in from secondary storage, and whenever a reference to a virtual object is received as a parameter of a message and deserialized.
+
+The `bodyMaker` function is passed the virtual object's state as a parameter, and is expected to return a new Javascript object with methods that close over this state.
+
+One distinguished method, named `initialize`, is called when new instances are first created, and is expected to initialize the state object.  Any parameters passed to the maker function returned by `makeKind` are passed directly to the `initialize` method.
+
+For example:
+
+```javascript
+function counterBody(state) {
+  return {
+    initialize(name = 'thing') {
+      state.counter = 0;
+      state.name = name;
+    },
+    inc() {
+      state.counter += 1;
+    },
+    dec() {
+      state.counter -= 1;
+    },
+    reset() {
+      state.counter = 0;
+    },
+    rename(newName) {
+      state.name = newName;
+    },
+    getCount() {
+      return state.counter;
+    },
+    getName() {
+      return state.name;
+    },
+  };
+}
```

Comment:
```suggestion
function makeCounterKit(state) {
  const init = (name = 'thing') => {
    state.counter = 0;
    state.name = name;
  };
  const body = {
      inc() {
      state.counter += 1;
    },
    dec() {
      state.counter -= 1;
    },
    reset() {
      state.counter = 0;
    },
    rename(newName) {
      state.name = newName;
    },
    getCount() {
      return state.counter;
    },
    getName() {
      return state.name;
    },
  };
  return harden({
    init,
    body,
  });
}
```
`makeCounterKit` makes a counter kit, which is a record with an `init` function and a `body` object. You can write it as above, which also gives a name to the `body` object that the `init` function` and `body` methods can use. Since none of these do, this one could be written with everything inline

```js
function makeCounterKit(state) {
  return harden({
    init(name = 'thing') {
      state.counter = 0;
      state.name = name;
    },
    body: {
      inc() {
        state.counter += 1;
      },
      dec() {
        state.counter -= 1;
      },
      reset() {
        state.counter = 0;
      },
      rename(newName) {
        state.name = newName;
      },
      getCount() {
        return state.counter;
      },
      getName() {
        return state.name;
      },
    },
  });
}
```

Both of these harden before anything happens, protecting ourselves from undesired mutation.

Reply:
The kit properties should probably be named `{init, self}` rather than `{init, body}`.

Reply:
This is still too complicated and ugly to my eyes and full of places to make mistakes.

Would you feel better if instead of deleting `initialize` we just replaced it with `()=>{}`?

BTW, this example illustrates an interesting difference between our respective API intuitions. I tend to see any place where the user needs to add a `harden` before passing to our SDK as an ergonomics problem.  Basically, if they need to put in a `harden` to be safe and they don't, then they've unwittingly made themselves vulnerable.  If the context is such that the harden should always be there, then the SDK should do it for them.  (Better would be a language where every API surface is hardened at birth, but that's not the language we have.)

Reply:
> Would you feel better if instead of deleting initialize we just replaced it with ()=>{}?

Worse. This goes in the wrong direction. Digs the hole deeper.

> Better would be a language where every API surface is hardened at birth, but that's not the language we have

Agreed on that. I also agree it is not clear how to best recover from this mistake. But the recovery should not require subtle reasoning on the part of the programmer to stay safe.

Reply:
Code should say what it means. In that regard, my code is *much* simpler than your's because they both describe an object with no `initialize` method but only my code looks like it does that. Your code is smaller but misleads.